### PR TITLE
Better error diagnostics and other fixes for extension methods

### LINKED ISSUE #9507: Inconsistency with partial application/currying between extension methods and functions
## Minimized example

```Scala
extension (x: Int)
  def add(y: Int): Int = x + y

def newFunction: Int => Int = 4.add

object Main extends App:
  println(newFunction(1))
```

## Output

compiler error:
```scala
value add is not a member of Int
An extension method was tried, but could not be fully constructed:
  {
    def $anonfun(y: Int): Int = extension_add(4)(y)
    closure($anonfun)
  }
```

## Expectation
`5`

## Explanation

There are 3 other similar cases in which this works as expected, so the above example seems like the odd one out.

1.
```scala
extension (x: Int)
  def add: Int => Int = y => x + y

def newFunction: Int => Int = 4.add
```
2.
```scala
def add: Int => Int = y => 4 + y

def newFunction: Int => Int = add
```
3.
```scala
def add(x: Int): Int = x + 4

def newFunction: Int => Int = add
```

The first two are similar in that both `add` functions have a return type of `Int => Int`.

The interesting bit is that the third one  works as expected (converting a function defined with `(x: Int): Int` into `Int => Int`, but this same conversion doesn't take place inside the context of an extension.

I don't know how extensions are actually implemented, and this might be expected behavior, which is why I didn't categorize this as a bug, just figured I'd mention since it seemed odd to me.

### LINKED ISSUE #10391: Extension methods don't get eta-expanded
## Minimized code

```Scala
extension (s: String) def x(i: Int): Int = i

val x = "".x
```

## Output

```
value x is not a member of String.
An extension method was tried, but could not be fully constructed:
    {
      def $anonfun(i: Int): Int = x("")(i)
      closure($anonfun)
    }
```

## Expectation

`"".x` should evaluate into a lambda just like it works for old style extensions using implicit classes:

```scala
implicit class StrOps(s: String) {
  def x(i: Int): Int = i
}

val x = "".x
```

results with

```scala
val x: Int => Int = Lambda$1578/0x00000008011a0040@71897380
```

### LINKED ISSUE #10810: Command line option aliases ("abbreviations") not understood by scalac
The aliases ("abbreviations") given to command line options in `dotty/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala` by `withAbbreviation` do not appear to be understood when specified on the command line.

I believe this is the cause of the community build test failures in #10733.

This also means that any such aliases added for backward compatibility as part of the recent compiler flags cleanup (#10236) may not be functional.

A couple of examples:

https://github.com/lampepfl/dotty/blob/bbd092b7f6ceeb7959e304a16219a040a4db1001/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala#L27

```
$ scalac -version
Scala compiler version 3.0.0-M3-bin-SNAPSHOT-git-bbd092b -- Copyright 2002-2020, LAMP/EPFL

$ scalac --version
bad option '--version' was ignored
Usage: scalac <options> <source files>
      ... help text elided ...
```
---

https://github.com/lampepfl/dotty/blob/bbd092b7f6ceeb7959e304a16219a040a4db1001/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala#L103

```
$ ls -l
total 8
-rw-r--r-- 1 tgrigg tgrigg 282 Dec 15 13:25 Test.class
-rw-r--r-- 1 tgrigg tgrigg 259 Dec 15 13:25 Test.tasty

$ dotc -from-tasty Test.tasty
$ ls -l
total 8
-rw-r--r-- 1 tgrigg tgrigg 282 Dec 15 13:26 Test.class
-rw-r--r-- 1 tgrigg tgrigg 259 Dec 15 13:26 Test.tasty

$ dotc --from-tasty Test.tasty
bad option '--from-tasty' was ignored
-- [E080] Syntax Error: Test.tasty:1:0 -----------------------------------------
1 |\�����]������������ASTs�<empty>�Test�<init>�java�lang����Object���?����Unit�scala�SourceFile�annotation����internal�������String���?�����Test.scala�Positions�Comments����6��������p�Uu�6����)u�6� ��u�6���p�U3�@��������������ؠƆ���
  |^^^
  |Expected a toplevel definition
-- Error: Test.tasty:1:6 -------------------------------------------------------
1 |\�����]������������ASTs�<empty>�Test�<init>�java�lang����Object���?����Unit�scala�SourceFile�annotation����internal�������String���?�����Test.scala�Positions�Comments����6��������p�Uu�6����)u�6� ��u�6���p�U3�@��������������ؠƆ���
  |      ^
  |      illegal character '\u0000'

    ... more errors elided ...
```

### LINKED ISSUE #10901: Extension method tried, but could not be fully constructed when using polymorphic operand
## Minimized code

Possible bug in self-contained example below due to polymorphic types.

```Scala
import scala.annotation.targetName

object BugExp4Point2D {

  sealed trait ColumnType[T]
  case object DoubleT extends ColumnType[Double]
  case object IntT extends ColumnType[Int]

  object dsl {


    extension [T1:Numeric, T2:Numeric](x: T1)

      // N - N
      @targetName("point2DConstant")
      def º(y: T2): Point2D[T1,T2] = ???


      // N - C
      @targetName("point2DConstantData")
      def º(y: ColumnType[T2]): Point2D[T1,T2] = ???



    extension [T1:Numeric, T2:Numeric](x: ColumnType[T1])
      // C - C
      @targetName("point2DData")
      def º(y: ColumnType[T2]): Point2D[T1,T2] = ???

      // C - N
      @targetName("point2DDataConstant")
      def º(y: T2): Point2D[T1,T2] = ???


  }

  case class Point2D[T1:Numeric, T2:Numeric](x:T1, y:T2)

  import dsl._

  def main(args: Array[String]): Unit = {
    val x = IntT
    val y = DoubleT

    val pos1: Point2D[Int,Double] = x º y
    val pos2: Point2D[Int,Double] = 100 º 200.1
    val pos3: Point2D[Int,Double] = 101 º y
    val pos4: Point2D[Int,Double] = x º 201.1

  }
}
```

## Output

```scala
[error] -- [E008] Not Found Error: /home/hhh/IdeaProjects/snol/splotly/src/gg/BugExp4Point2D.scala:55:38 
[error] 55 |    val pos1: Point2D[Int,Double] = x º y
[error]    |                                    ^^^
[error]    |      value º is not a member of object gg.BugExp4Point2D.IntT.
[error]    |      An extension method was tried, but could not be fully constructed:
[error]    |
[error]    |          º(x)
[error] -- [E008] Not Found Error: /home/hhh/IdeaProjects/snol/splotly/src/gg/BugExp4Point2D.scala:58:38 
[error] 58 |    val pos4: Point2D[Int,Double] = x º 201.1
[error]    |                                    ^^^
[error]    |      value º is not a member of object gg.BugExp4Point2D.IntT.
[error]    |      An extension method was tried, but could not be fully constructed:
[error]    |
[error]    |          º(x)
```

## Expectation

I expected `pos1` and `pos4` to be handled correctly just as `pos2` and `pos3` are.
What is different in `pos1` and `pos4` is that the first operand uses a polymorphic type parameter.
If it is a bug, any way of circumventing this temporarily? 

Tested on M3.

EDIT: Forgot to indicate that if we remove the last extension and `pos4`, the example compiles correctly.

### PR DESCRIPTION
**Pull request by odersky**
We gave misleading diagnostic before in the case where no extension
method was found since a TypeError was thrown.

Also: Avoid redundant test that prevented eta expansion.
Also: Drop support for extension_ names.

Fixes #10391 
Fixes #9507
Fixes #10901

### CODE REVIEW

File: tests/neg/i10810.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+final case class A()
+final case class B(a:A)
+
+object Test:
```
Comment (bishabosha) [inline tests/neg/i10810.scala:4]: This file and commit reference the wrong issue, #10810 is for command line abbreviations

Comment (odersky) [reply tests/neg/i10810.scala]: Right. It was #10870.

**Comment by bishabosha** [comment]
with relation to #10044 , the error now becomes a suggestion to `import opaques.arrayOps`:
```scala
scala> IArray(1,2,3).reverse.sorted                                                                                   
val res0: opaques.IArray[Int] = Array(1, 2, 3)

scala> IArray(1,2,3).reverse.sorted
1 |IArray(1,2,3).reverse.sorted
  |^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |value sorted is not a member of scala.opaques.IArray[Int], but could be made available as an extension method.
  |
  |The following import might fix the problem:
  |
  |  import opaques.arrayOps
  |

```

**Review by bishabosha — CHANGES_REQUESTED** [review]
code like the issue #7056 no longer give a suggestion of a tried extension method, and instead suggest that nothing was found, e.g.
```scala
type A
type B <: A

type PartialId[X] = X match {
  case B => X
}

trait T1[T] {
  extension (t1: T) def idnt1: Any
}

given [T <: A](using PartialId[T]): T1[T] = new T1[T] {
  extension (t1: T) def idnt1: Any = ???
}

given PartialId[B] = ???

val x: B = ???
val z = x.idnt1
```
errors with the following:
```scala
-- [E008] Not Found Error: sandbox/cls/example.scala:19:10 ---------------------
19 |val z = x.idnt1
   |        ^^^^^^^
   |        value idnt1 is not a member of B
1 error found
```
which I believe is probably less helpful compared to the error in M3:
```scala
-- [E008] Not Found Error: example.scala:19:10 ---------------------------------
19 |val z = x.idnt1
   |        ^^^^^^^
   |      value idnt1 is not a member of B.
   |      An extension method was tried, but could not be fully constructed:
   |
   |          example$package.given_T1_T[T](given_PartialId_B).idnt1()
1 error found

```

**Comment by odersky** [comment]
The addendums for the implicit search errors were dropped by accident. Last commit brings them back. 

The error message for #10044 is also the same as it was.

**Pull request merged**